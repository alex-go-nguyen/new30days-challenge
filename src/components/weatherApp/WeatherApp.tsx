import React, { useEffect, useState } from 'react';
import classNames from 'classnames/bind';
import styles from './WeatherApp.module.scss';
import { FaCloud, FaRegEye, FaWind } from 'react-icons/fa';
import axios from 'axios';
import { useForm } from 'react-hook-form';
const cx = classNames.bind(styles);

type data = {
    city: string;
    country: string;
    weather: string;
    deg: string;
    time: string;
    visibility: string;
    wind: string;
    cloud: string;
};

type UserSubmitForm = {
    city: string;
};

export interface WeatherAppProps {
    data: data;
}

export default function WeatherApp({ data }: WeatherAppProps) {
    const { register, handleSubmit, reset } = useForm<UserSubmitForm>();
    const [dataWeather, setDataWeather] = useState(data);
    const CallAPI = async (place: string) => {
        try {
            const { data } = await axios.get('https://v1.nocodeapi.com/danish/ow/kByYHeRdQCmpeBOG/byCityName', {
                params: {
                    q: place,
                },
            });
            const currentDate = new Date();
            console.log(data);
            setDataWeather({
                city: data.name,
                country: data.sys.country,
                weather: data.weather[0].main,
                deg: (data.main.temp - 273.15).toFixed(0).toString(),
                time: currentDate.toLocaleDateString() + ', ' + currentDate.toLocaleTimeString(),
                visibility: data.visibility,
                wind: data.wind.speed,
                cloud: data.clouds.all,
            });
        } catch (e) {}
    };
    useEffect(() => {
        CallAPI('london');
    }, []);

    const onSubmitHandler = (data: UserSubmitForm) => {
        CallAPI(data.city);
        reset();
    };
    return (
        <form className={cx('container')} onSubmit={handleSubmit(onSubmitHandler)}>
            <style jsx>
                {`
                    form {
                        background-image: url('${parseInt(dataWeather?.deg) > 15
                            ? 'https://raw.githubusercontent.com/namndwebdev/html-css-js-thuc-chien/main/Weather%20App/hot.png'
                            : 'https://raw.githubusercontent.com/namndwebdev/html-css-js-thuc-chien/main/Weather%20App/cold.png'}');
                    }
                `}
            </style>
            <input placeholder="Search..." className={cx('search')} {...register('city')} />
            <div className={cx('body')}>
                <p className={cx('place')}>
                    {dataWeather?.city} , {dataWeather?.country}
                </p>
                <p className={cx('time')}>{dataWeather?.time}</p>
                <div className={cx('temperature')}>
                    <span className={cx('value')}>{dataWeather?.deg}</span>
                    <span className={cx('type')}>C</span>
                </div>
                <p className={cx('short-desc')}>{dataWeather?.weather}</p>
                <div className={cx('more-desc')}>
                    <div className={cx('visibility')}>
                        <div className={cx('icon')}>
                            <FaRegEye />
                        </div>
                        <span>{dataWeather?.visibility} (m)</span>
                    </div>
                    <div className={cx('wind')}>
                        <div className={cx('icon')}>
                            <FaWind />
                        </div>
                        <span>{dataWeather?.wind} (m/s)</span>
                    </div>
                    <div className={cx('cloud')}>
                        <div className={cx('icon')}>
                            <FaCloud />
                        </div>
                        <span>{dataWeather?.cloud} (%)</span>
                    </div>
                </div>
            </div>
        </form>
    );
}
