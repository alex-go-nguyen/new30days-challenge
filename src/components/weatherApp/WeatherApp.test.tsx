import React from 'react';
import renderer, { ReactTestRenderer, ReactTestRendererJSON, act } from 'react-test-renderer';
import WeatherApp from './WeatherApp';

describe('WeatherApp component', () => {
    const component = renderer.create(
        <WeatherApp
            data={{
                city: 'string',
                country: 'string',
                weather: 'string',
                deg: 'string',
                time: 'string',
                visibility: 'string',
                wind: 'string',
                cloud: 'string',
            }}
        />,
    );
    let tree = component.toJSON() as ReactTestRendererJSON;
    test('should render correctly', () => {
        expect(tree).toMatchSnapshot();
    });
});
