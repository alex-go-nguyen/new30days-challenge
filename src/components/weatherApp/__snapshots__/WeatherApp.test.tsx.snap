// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherApp component should render correctly 1`] = `
<form
  autoComplete="off"
  className="container"
  onSubmit={[Function]}
>
  <style
    jsx={true}
  >
    
                    form {
                        background-image: url('https://raw.githubusercontent.com/namndwebdev/html-css-js-thuc-chien/main/Weather%20App/cold.png');
                    }
                
  </style>
  <input
    autoComplete="off"
    className="search"
    name="city"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Search..."
  />
  <div
    className="body"
  >
    <p
      className="place"
    >
      string
       , 
      string
    </p>
    <p
      className="time"
    >
      string
    </p>
    <div
      className="temperature"
    >
      <span
        className="value"
      >
        string
      </span>
      <span
        className="type"
      >
        C
      </span>
    </div>
    <p
      className="short-desc"
    >
      string
    </p>
    <div
      className="more-desc"
    >
      <div
        className="visibility"
      >
        <div
          className="icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M288 144a110.94 110.94 0 0 0-31.24 5 55.4 55.4 0 0 1 7.24 27 56 56 0 0 1-56 56 55.4 55.4 0 0 1-27-7.24A111.71 111.71 0 1 0 288 144zm284.52 97.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400c-98.65 0-189.09-55-237.93-144C98.91 167 189.34 112 288 112s189.09 55 237.93 144C477.1 345 386.66 400 288 400z"
            />
          </svg>
        </div>
        <span>
          string
           (m)
        </span>
      </div>
      <div
        className="wind"
      >
        <div
          className="icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M156.7 256H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h142.2c15.9 0 30.8 10.9 33.4 26.6 3.3 20-12.1 37.4-31.6 37.4-14.1 0-26.1-9.2-30.4-21.9-2.1-6.3-8.6-10.1-15.2-10.1H81.6c-9.8 0-17.7 8.8-15.9 18.4 8.6 44.1 47.6 77.6 94.2 77.6 57.1 0 102.7-50.1 95.2-108.6C249 291 205.4 256 156.7 256zM16 224h336c59.7 0 106.8-54.8 93.8-116.7-7.6-36.2-36.9-65.5-73.1-73.1-55.4-11.6-105.1 24.9-114.9 75.5-1.9 9.6 6.1 18.3 15.8 18.3h32.8c6.7 0 13.1-3.8 15.2-10.1C325.9 105.2 337.9 96 352 96c19.4 0 34.9 17.4 31.6 37.4-2.6 15.7-17.4 26.6-33.4 26.6H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16zm384 32H243.7c19.3 16.6 33.2 38.8 39.8 64H400c26.5 0 48 21.5 48 48s-21.5 48-48 48c-17.9 0-33.3-9.9-41.6-24.4-2.9-5-8.7-7.6-14.5-7.6h-33.8c-10.9 0-19 10.8-15.3 21.1 17.8 50.6 70.5 84.8 129.4 72.3 41.2-8.7 75.1-41.6 84.7-82.7C526 321.5 470.5 256 400 256z"
            />
          </svg>
        </div>
        <span>
          string
           (m/s)
        </span>
      </div>
      <div
        className="cloud"
      >
        <div
          className="icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M537.6 226.6c4.1-10.7 6.4-22.4 6.4-34.6 0-53-43-96-96-96-19.7 0-38.1 6-53.3 16.2C367 64.2 315.3 32 256 32c-88.4 0-160 71.6-160 160 0 2.7.1 5.4.2 8.1C40.2 219.8 0 273.2 0 336c0 79.5 64.5 144 144 144h368c70.7 0 128-57.3 128-128 0-61.9-44-113.6-102.4-125.4z"
            />
          </svg>
        </div>
        <span>
          string
           (%)
        </span>
      </div>
    </div>
  </div>
</form>
`;
